### PR TITLE
Make the read call control the actual read size in KfsInputChannel, if read ahead logic is turned off

### PR DESCRIPTION
**Pull request by mckurt**
Currently, QFS access java API reads a default 1MB from a file, each time it calls the corresponding JNI read function. The proposed change modifies this logic. Accordingly, if the user disables the read ahead logic (by assigning 0 to it), the amount of data read from QFS C++ client never exceeds the amount of data requested by the user.

### CODE REVIEW

File: src/java/qfs-access/src/main/java/com/quantcast/qfs/access/BufferPool.java [hunk 1]
```diff
@@ -28,10 +28,11 @@
 import java.nio.ByteBuffer;
 import java.util.AbstractQueue;
 import java.util.concurrent.ConcurrentLinkedQueue;
+import java.lang.Math;
```
Comment (cwoerner) [inline src/java/qfs-access/src/main/java/com/quantcast/qfs/access/BufferPool.java]: Why import java.lang.Math?

Comment (mckurt) [inline src/java/qfs-access/src/main/java/com/quantcast/qfs/access/BufferPool.java]: Fixed. Thanks.

File: src/java/qfs-access/src/main/java/com/quantcast/qfs/access/KfsInputChannel.java [hunk 2]
```diff
@@ -197,6 +200,7 @@ protected void finalize() throws Throwable
     public void setReadAheadSize(long readAheadSize) {
         if(readAheadSize >= 0) {
             kfsAccess.kfs_setReadAheadSize(kfsFd, readAheadSize);
+            isReadAheadOff = (readAheadSize == 0 ? true : false);
```
Comment (mikeov) [inline src/java/qfs-access/src/main/java/com/quantcast/qfs/access/KfsInputChannel.java]: isReadAheadOff = 0 == readAheadSize;

File: src/java/qfs-access/src/main/java/com/quantcast/qfs/access/KfsInputChannel.java [hunk 3]
```diff
@@ -114,18 +115,20 @@ public synchronized int read(ByteBuffer dst) throws IOException
...
 
-    private void readDirect(ByteBuffer buf) throws IOException
+    private void readDirect(ByteBuffer buf, int remRequestedBytes) throws IOException
     {
         if (!buf.isDirect()) {
             throw new IllegalArgumentException("need direct buffer");
         }
         final int pos = buf.position();
-        final int sz  = read(kfsAccess.getCPtr(), kfsFd, buf, pos, buf.limit());
+        final int end = (isReadAheadOff && remRequestedBytes > 0 ?
```
Comment (mikeov) [inline src/java/qfs-access/src/main/java/com/quantcast/qfs/access/KfsInputChannel.java]: final int end = (isReadAheadOff && remRequestedBytes > 0) ?
                 Math.min(buf.limit(), pos + remRequestedBytes) : buf.limit();

File: src/java/qfs-access/src/main/java/com/quantcast/qfs/access/KfsTest.java [hunk 4]
```diff
@@ -503,4 +506,46 @@ private static void verifyFileAttr(KfsAccess kfsAccess,
...
+
+        KfsInputChannel inputChannel = kfsAccess.kfs_open(filePath);
+        if (inputChannel == null) {
+            System.out.println("open on " + filePath + "failed!");
+            System.exit(1);
+        }
+        inputChannel.setReadAheadSize(0);
+
+        byte[] dstBuf = new byte[128];
+        res = inputChannel.read(ByteBuffer.wrap(dstBuf, 0, 128));
```
Comment (mikeov) [inline src/java/qfs-access/src/main/java/com/quantcast/qfs/access/KfsTest.java]: Issuing single read doesn't seem that interesting. Better cover coverage at low cost: small read seek forward (skip), and possible seek back.

File: src/java/qfs-access/src/main/java/com/quantcast/qfs/access/KfsInputChannel.java (hunk 2 above)
Comment (mckurt) [inline src/java/qfs-access/src/main/java/com/quantcast/qfs/access/KfsInputChannel.java]: fixed.

File: src/java/qfs-access/src/main/java/com/quantcast/qfs/access/KfsInputChannel.java (hunk 3 above)
Comment (mckurt) [inline src/java/qfs-access/src/main/java/com/quantcast/qfs/access/KfsInputChannel.java]: fixed.

File: src/java/qfs-access/src/main/java/com/quantcast/qfs/access/KfsTest.java (hunk 4 above)
Comment (mckurt) [inline src/java/qfs-access/src/main/java/com/quantcast/qfs/access/KfsTest.java]: added these;
-seek forward
-do small read
-seek back
-enable read ahead
-do small read

**Pull request merged**